src/theory/sets/theory_sets_rels.cpp

```diff
@@ -1183,14 +1185,15 @@ void TheorySetsRels::check(Theory::Effort level)
     return false;
   }
 
-  void TheorySetsRels::makeSharedTerm( Node n ) {
+  void TheorySetsRels::makeSharedTerm(Node n, TypeNode t)
+  {
     if (d_shared_terms.find(n) != d_shared_terms.end())
     {
       return;
     }
     Trace("rels-share") << " [sets-rels] making shared term " << n << std::endl;
     // force a proxy lemma to be sent for the singleton containing n
-    Node ss = NodeManager::currentNM()->mkNode(SINGLETON, n);
+    Node ss = NodeManager::currentNM()->mkSingleton(t, n);
```

Comment:
It is not possible to figure out if this singleton is safe without knowing the type. So the function is updated to receive the type as an argument. See the justification for the types at the locations where `makeSharedTerm(Node n, TypeNode t)` is called.